utils/sample_qc.py

```diff
@@ -0,0 +1,487 @@
+
+    strata_exp = hl.tuple([ht[x] for x in strata]) if strata else hl.tuple([])
+
+    fail_exprs = {
+        f'fail_{metric}':
+            (ht[metric] >= ht.metrics_stats[strata_exp][metric].upper) |
+            (ht[metric] <= ht.metrics_stats[strata_exp][metric].lower)
+        for metric in qc_metrics}
+    ht = ht.transmute(**fail_exprs)
+    pop_platform_filters = make_pop_filters_expr(ht, qc_metrics)
+    return ht.annotate(pop_platform_filters=pop_platform_filters)
+
+
+def flatten_duplicate_samples_ht(dups_ht: hl.Table) -> hl.Table:
+    """
+    Flattens the result of `filter_duplicate_samples`, so that each line contains a single sample.
+    An additional annotation is added: `dup_filtered` indicating which of the duplicated samples was kept.
+
+    :param Table dups_ht: Input HT
+    :return: Flattened HT
+    :rtype: Table
+    """
+    dups_ht = dups_ht.key_by()
+    dups_ht = dups_ht.annotate(
+        dups=hl.array([(dups_ht.s_kept, True)]).extend(
+            dups_ht.filtered.map(lambda x: (x.s, False))
+        )
+    )
+    dups_ht = dups_ht.explode('dups')
+    return dups_ht.select(s=dups_ht.dups[0], dup_filtered=dups_ht.dups[1]).key_by('s')
+
+
+def add_filters_expr(
+        filters: Dict[str, hl.expr.BooleanExpression],
+        current_filters: hl.expr.SetExpression = None
+) -> hl.expr.SetExpression:
+    """
+    Creates an expression to create or add filters.
+    For each entry in the `filters` dictionary, if the value evaluates to `True`,
+    then the key is added as a filter name.
+
+    Current filters are kept if provided using `current_filters`
+
+    :param dict of str -> BooleanExpression filters: The filters and their expressions
+    :param SetExpression current_filters: The set of current filters
+    :return: An expression that can be used to annotate the filters
+    :rtype: SetExpression
+    """
+    if current_filters is None:
+        current_filters = hl.empty_set(hl.tstr)
+
+    return hl.fold(
+            lambda x, y: x.union(y),
+            current_filters,
+            [
+                hl.cond(filter_condition, hl.set([filter_name]), hl.empty_set(hl.tstr))
+                for filter_name, filter_condition in filters.items()
+            ]
+        )
+
+
+def get_platform_specific_intervals(platform_pc_loadings_ht: hl.Table, threshold: float) -> List[hl.Interval]:
+    """
+    This takes the platform PC loadings returns a list of intervals where the sum of the loadings above the given threshold.
+    The experimental / untested idea behind this, is that those intervals may be problematic on some platforms.
+
+    :param Table platform_pc_loadings_ht: Platform PCA loadings
+    :param float threshold: Minimal threshold
+    :return: List of intervals with PC loadings above the given threshold
+    :rtype: list of Interval
+    """
+    intervals = hl.import_locus_intervals(exome_calling_intervals_path)
+    intervals = intervals.key_by('target')
```

Comment:
Same comment as above

Reply:
So, I actually believe that we should key_by interval rather than `target`, I need to test this but I think it'd be best (and we can keep whatever other annotations like target are present)

Reply:
I've now changed things so, there is no more hard-coded things and it's keyed by interval!